crates/ika-core/src/dwallet_mpc/crytographic_computation/mpc_computations/presign.rs

```diff
@@ -4,66 +4,306 @@
+                        access_structure,
+                        consensus_round,
+                        &serialized_messages_by_consensus_round,
+                    )?;
+
+                advance_request.map(PresignAdvanceRequestByProtocol::Secp256r1ECDSA)
             }
         };
+        Ok(advance_request)
+    }
+}
+
+impl PresignPublicInputByProtocol {
+    pub(crate) fn try_new(
+        protocol: DWalletSignatureScheme,
+        versioned_network_encryption_key_public_data: &VersionedNetworkEncryptionKeyPublicData,
+        dwallet_public_output: Option<SerializedWrappedMPCPublicOutput>,
+    ) -> DwalletMPCResult<Self> {
+        // if dwallet_public_output is none, try v2, else if some deserialize and match version
+        if dwallet_public_output.is_none() {
+            return Self::try_new_v2(protocol, versioned_network_encryption_key_public_data, None);
+        }
+
+        // Safe to unwrap as we checked for None above
+        match bcs::from_bytes(&dwallet_public_output.unwrap())? {
+            VersionedDwalletDKGSecondRoundPublicOutput::V1(dkg_output) => {
+                Self::try_new_v1(versioned_network_encryption_key_public_data, dkg_output)
+            }
+            VersionedDwalletDKGSecondRoundPublicOutput::V2(dkg_output) => Self::try_new_v2(
+                protocol,
+                versioned_network_encryption_key_public_data,
+                Some(dkg_output),
+            ),
+        }
+    }
+    pub(crate) fn try_new_v1(
+        versioned_network_encryption_key_public_data: &VersionedNetworkEncryptionKeyPublicData,
+        dwallet_public_output: MPCPublicOutput,
+    ) -> DwalletMPCResult<Self> {
+        let decentralized_party_dkg_output = bcs::from_bytes::<
+            <Secp256K1ECDSAProtocol as dkg::Protocol>::DecentralizedPartyTargetedDKGOutput,
+        >(&dwallet_public_output)?
+        .into();
 
-        let pub_input = Self::PublicInput {
-            protocol_public_parameters,
-            dkg_output: decentralized_dkg_output,
+        let protocol_public_parameters =
+            versioned_network_encryption_key_public_data.secp256k1_protocol_public_parameters();
+
+        let public_input: <PresignParty<Secp256K1ECDSAProtocol> as mpc::Party>::PublicInput =
+            (protocol_public_parameters, decentralized_party_dkg_output).into();
+        Ok(PresignPublicInputByProtocol::Secp256k1ECDSA(public_input))
+    }
+
+    pub(crate) fn try_new_v2(
+        protocol: DWalletSignatureScheme,
+        versioned_network_encryption_key_public_data: &VersionedNetworkEncryptionKeyPublicData,
+        dwallet_public_output: Option<MPCPublicOutput>,
+    ) -> DwalletMPCResult<Self> {
+        let input = match protocol {
+            DWalletSignatureScheme::ECDSASecp256k1 => {
+                let protocol_public_parameters = versioned_network_encryption_key_public_data
+                    .secp256k1_protocol_public_parameters();
+
+                let deserialized_dwallet_public_output =
```

Comment:
which type are you deserializing into? be specific. It should be the `DecentralizedPartyDKGOutput`. This you should later convert by `into()` into `DecentralizedPartyTargetedDKGOutput`.